CQLCookbook.md

```diff
@@ -0,0 +1,349 @@
+      value: Qty.value.value,
+      unit: Coalesce(Qty.unit.value, Qty.code.value)
+    }
+  else null
+```
+Variables:
+- *Observation_Codes_Or_Valueset:* Replace with the name of a valueset of Observations or a defined group of Observation codes that has been established in a previous define statement.
+- *time_interval:* How far back from today you want to look in the patient's EHR for the Observation.
+- *unit:* Unit of measurement for lab result (ex. mm[Hg], %, pH).
+
+Example Implementation: [HomeOxygenTherapyPrepopulation-0.1.0.cql](https://github.com/HL7-DaVinci/CDS-Library/blob/master/HomeOxygenTherapy/R4/files/HomeOxygenTherapyPrepopulation-0.1.0.cql)
+
+### Extract Performer Field Information from an Observation
+The 'performer' field has a value of an array with references. Extract the display name of a reference within an Observation. Examples of this can be a reference to a Practitioner who performed the observation or the Organization where the Observation was performed.
+```sql
+// Retrieve the 'performer' field of the first observation in the lab.
+define "Lab Performer":
+  First([Observation: "My_Lab"]).performer
+  
+// Extract the display value of the Practitioner
+define "Tester":
+  if exists("Lab Performer") then 
+    "Lab" P
+    where P.type = 'Practitioner'
+    return P.display.value
+  else
+    null  
+    
+// Extract the display value of the Organization
+define "TestLaboratory":
+  if exists("Lab Performer") then 
+    "Lab" P
+    where P.type = 'Organization'
+    return P.display.value
+  else
+    null  
+```
+Variables:
+- *My_Lab:* Name of a list of lab Observations or an Observation value set.
+
+Example Implementation: [RespiratoryAssistDevicesPrepopulation-0.1.0.cql](https://github.com/HL7-DaVinci/CDS-Library/blob/master/RespiratoryAssistDevices/R4/files/RespiratoryAssistDevicesPrepopulation-0.1.0.cql)
+
+***
+
+## 3) Example Prepopulation.cql file
+Below is a full example of a ruleset prepopulation file. This file prepopulates the questionnaire for the Non Emergency Ambulance Transportation ruleset, and is stored in [CDS-Library/NonEmergencyAmbulanceTransportation/R4/files/NonEmergencyAmbulanceTransportationPrepopulation-0.1.0.cql](https://github.com/HL7-DaVinci/CDS-Library/blob/Shared_CQL_Library/NonEmergencyAmbulanceTransportation/R4/files/NonEmergencyAmbulanceTransportationPrepopulation-0.1.0.cql).
+```sql
+library NonEmergencyAmbulanceTransportationPrepopulation  version '0.1.0'
+using FHIR version '4.0.0'
+include FHIRHelpers version '4.0.0' called FHIRHelpers
+
+valueset "Ambulance Transport Procedure": 'http://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113762.1.4.1219.108'
+
+parameter service_request ServiceRequest
+
+context Patient
+
+define "ServiceRequestHcpcsCoding": singleton from (
+  ((cast service_request.code as CodeableConcept).coding) coding
+    where coding.system.value = 'https://bluebutton.cms.gov/resources/codesystem/hcpcs'
+    return FHIRHelpers.ToCode(coding))
```

Comment:
Is it worth mentioning the FHIRHelpters a little bit as they are used in the code? An example to include other libraries and use them.

Reply:
Great point. I added a note about it in the "Header" section. I also added instruction on how to actually call external library functions within the "Function Statements" section